crypto/defaults.c

```diff
@@ -0,0 +1,200 @@
+ */
+
+static char enginesdir[MAX_PATH + 1];
+
+/**
+ * @brief The pointer to the enginesdir buffer
+ */
+static char *enginesdirptr = NULL;
+
+/**
+ * @brief The directory where OpenSSL modules are located.
+ */
+static char modulesdir[MAX_PATH + 1];
+
+/**
+ * @brief The pointer to the modulesdir buffer
+ */
+static char *modulesdirptr = NULL;
+
+/**
+ * @brief Get the list of Windows registry directories.
+ *
+ * This function retrieves a list of Windows registry directories.
+ *
+ * @return A pointer to a char array containing the registry directories.
+ */
+static char *get_windows_regdirs(char *dst, LPCTSTR valuename)
+{
+    char *retval = NULL;
+#ifdef REGISTY_KEY
+    DWORD keysize;
+    DWORD ktype;
+    HKEY hkey;
+    LSTATUS ret;
+    DWORD index = 0;
+    LPCTCH tempstr = NULL;
+   
+    ret = RegOpenKeyEx(HKEY_LOCAL_MACHINE,
+                       TEXT(REGISTRY_KEY), KEY_WOW64_32KEY,
+                       KEY_QUERY_VALUE, &hkey);
+    if (ret != ERROR_SUCCESS)
+        goto out;
+
+    ret = RegQueryValueEx(hkey, valuename, NULL, &ktype, NULL,
+                          &keysize);
+    if (ret != ERROR_SUCCESS)
+        goto out;
+    if (ktype != REG_EXPAND_SZ)
+        goto out;
+    if (keysize > MAX_PATH)
+        goto out;
+
+    keysize++;
+    tempstr = OPENSSL_zalloc(keysize * sizeof(TCHAR));
+
+    if (tempstr == NULL)
+        goto out;
+
+    if (RegQueryValueEx(hkey, valuename,
+                        NULL, &ktype, tempstr, &keysize) != ERROR_SUCCESS)
+        goto out;
+
+    if (!WideCharToMultiByte(CP_UTF8, 0, tempstr, -1, dst, keysize,
+                             NULL, NULL)) 
+        goto out;
+
+    retval = dst;
+out:
+    OPENSSL_free(tempstr);
+    RegCloseKey(hkey);
+#endif
+    return retval;
+}
+
+static CRYPTO_ONCE defaults_setup_init = CRYPTO_ONCE_STATIC_INIT;
+
+/**
+ * @brief Function to setup default values to run once.
+ * Only used in Windows environments.  Does run time initalization
+ * of openssldir/modulesdir/enginesdir from the registry
+ */
+DEFINE_RUN_ONCE_STATIC(do_defaults_setup)
+{
+    get_windows_regdirs(openssldir, TEXT("OPENSSLDIR"));
+    get_windows_regdirs(enginesdir, TEXT("ENGINESDIR"));
+    get_windows_regdirs(modulesdir, TEXT("MODULESDIR"));
+
+    /*
+     * Set our pointers only if the directories are fetched properly
+     */
+    if (strlen(openssldir) > 0)
+        openssldirptr = openssldir;
+
+    if (strlen(enginesdir) > 0)
+        enginesdirptr = enginesdir;
+
+    if (strlen(modulesdir) > 0)
+        modulesdirptr = modulesdir;
+
+    return 1;
+}
+#endif
+
+/**
+ * @brief Get the directory where OpenSSL is installed.
+ *
+ * @return A pointer to a string containing the OpenSSL directory path.
+ */
+const char *ossl_get_openssldir(void)
+{
+#if defined(_WIN32) && defined (OSSL_WINCTX)
+    if (!RUN_ONCE(&defaults_setup_init, do_defaults_setup))
+        return NULL;
+    return (const char *)openssldirptr;
+# else
+    return OPENSSLDIR;
```

Comment:
In the docs you said no attempt is made to load if OSSL_WINCTX is not defined, but this seems to just fallback to OPENSSLDIR? That's probably the correct thing to do - otherwise this wouldn't be backwards compatible

Reply:
yes, correct in the table I'll be posting soon

Reply:
If "backward compatibility" with this vulnerability is desired, yes. Otherwise this should be fixed to never use a hard-coded `OPENSSLDIR` directory when `_WIN32`, because it's demonstrated to allow DLL injection by an unprivileged user on that plaform. Ref #24528.

This, complemented by setting `OSSL_WINCTX` to `openssl` by default would fix the problem.